src/NHibernate/Hql/Ast/ANTLR/HqlSqlWalker.cs

```diff
@@ -248,6 +248,14 @@ private IASTNode GenerateVersionPropertyNode(IQueryable persister)
 			return versionPropertyNode;
 		}
 
+		partial void EnterRule_insertStatement()
+		{
+			// Filters are not supported by DML, and they wreck the select part of an insert.
+			// Unfortunately I do not see any better place to handle this. This hack requires
+			// EnabledFilters to not be the same instance than the one held by the session.
+			_qti.EnabledFilters.Clear();
```

Comment:
This is hackish. Any better idea is welcome.

The trouble is, the query translator has no idea it is a DML query before having analysed it. But the analyse does already incorporate filters in the AST if any is enabled. So ideally we need to ignore filters before analysing the query if it is a DML, but we cannot know it is a DML before having analysed it.

In the analyse, the very first thing happening is finding the insert statement. So the hack is to clear the enabled filters at this point, in order to ignore them.

(We may instead set a new flag on the query translator for telling it very early it is a DML, and then the query translator could yield an empty `EnabledFilters` if it is a DML. It would avoid the dictionary copy, but overall the fix would still be a hack.)

Reply:
If we are thinking about backward compability we could add a default parameter to ExecuteUpdate?
ExecuteUpdate(bool enableFilters = false)

Reply:
The reason for this is explained [here](https://github.com/nhibernate/nhibernate-core/issues/1921#issuecomment-444941560).